lib/AST/DiagnosticEngine.cpp

```diff
@@ -1101,8 +1102,21 @@ DiagnosticEngine::diagnosticInfoForDiagnostic(const Diagnostic &diagnostic) {
     return None;
 
   // Figure out the source location.
-  SourceLoc loc = diagnostic.getLoc();
-  if (loc.isInvalid() && diagnostic.getDecl()) {
+  SourceLoc loc;
+  if (diagnostic.getLoc()) {
+    // Use the loc directly.
+    loc = diagnostic.getLoc();
+
+  } else if (diagnostic.getModuleFile()) {
+    // Reuse or generate a fake buffer holding the path to the loaded file.
+    const ModuleFile *file = diagnostic.getModuleFile();
+    auto filename = file->getModuleFilename();
+    auto bufferID = SourceMgr.getIDForBufferIdentifier(filename);
+    if (!bufferID)
+      bufferID = SourceMgr.addMemBufferCopy(filename, filename);
```

Comment:
Any particular reason to have the filename in the fake buffer? For debugging? Doesn't really hurt I suppose.

Reply:
Replaced it with an empty string.